### ISSUE TITLE
sphinx warnings `reference target not found`

### ISSUE THREAD

**Issue by kloczek** [python-gitlab/python-gitlab#1992]
On building my packages I'm using `sphinx-build` command with `-n` switch which shows warmings about missing references. These are not critical issues.
Here is the output with warnings:
```console
+ /usr/bin/sphinx-build -n -T -b man docs build/sphinx/man
Running Sphinx v4.5.0
making output directory... done
myst v0.17.2: MdParserConfig(commonmark_only=False, gfm_only=False, enable_extensions=[], linkify_fuzzy_links=True, dmath_allow_labels=True, dmath_allow_space=True, dmath_allow_digits=True, dmath_double_inline=False, update_mathjax=True, mathjax_classes='tex2jax_process|mathjax_process|math|output_area', disable_syntax=[], all_links_external=False, url_schemes=('http', 'https', 'mailto', 'ftp'), ref_domains=None, highlight_code_blocks=True, number_code_blocks=[], title_to_header=False, heading_anchors=None, heading_slug_func=None, footnote_transition=True, sub_delimiters=('{', '}'), words_per_minute=200)
[autosummary] generating autosummary for: api-objects.rst, api-usage.rst, api/gitlab.rst, api/gitlab.v4.rst, changelog.md, cli-examples.rst, cli-objects.rst, cli-usage.rst, faq.rst, gl_objects/access_requests.rst, ..., gl_objects/snippets.rst, gl_objects/system_hooks.rst, gl_objects/templates.rst, gl_objects/todos.rst, gl_objects/topics.rst, gl_objects/users.rst, gl_objects/variables.rst, gl_objects/wikis.rst, index.rst, release-notes.rst
building [mo]: targets for 0 po files that are out of date
building [man]: all manpages
updating environment: [new config] 65 added, 0 changed, 0 removed
reading sources... [100%] release-notes
looking for now-outdated files... none found
pickling environment... done
checking consistency... done
writing... python-gitlab.3 { cli-usage api-usage cli-examples api-objects gl_objects/access_requests gl_objects/appearance gl_objects/applications gl_objects/emojis gl_objects/badges gl_objects/branches gl_objects/clusters gl_objects/messages gl_objects/commits gl_objects/deploy_keys gl_objects/deploy_tokens gl_objects/deployments gl_objects/discussions gl_objects/environments gl_objects/events gl_objects/epics gl_objects/features gl_objects/geo_nodes gl_objects/groups gl_objects/group_access_tokens gl_objects/issues gl_objects/keys gl_objects/boards gl_objects/labels gl_objects/notifications gl_objects/merge_trains gl_objects/merge_requests gl_objects/merge_request_approvals gl_objects/milestones gl_objects/namespaces gl_objects/notes gl_objects/packages gl_objects/pagesdomains gl_objects/personal_access_tokens gl_objects/pipelines_and_jobs gl_objects/projects gl_objects/project_access_tokens gl_objects/protected_branches gl_objects/releases gl_objects/runners gl_objects/remote_mirrors gl_objects/repositories gl_objects/repository_tags gl_objects/search gl_objects/settings gl_objects/snippets gl_objects/system_hooks gl_objects/templates gl_objects/todos gl_objects/topics gl_objects/users gl_objects/variables gl_objects/sidekiq gl_objects/wikis api/gitlab api/gitlab.v4 cli-objects changelog release-notes faq } /home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/gl_objects/applications.rst:10: WARNING: py:class reference target not found: gitlab.v4.objects.Applications
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/gl_objects/epics.rst:15: WARNING: py:attr reference target not found: gitlab.Gitlab.Group.epics
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/gl_objects/epics.rst:54: WARNING: py:attr reference target not found: gitlab.Gitlab.GroupEpic.issues
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/gl_objects/group_access_tokens.rst:14: WARNING: py:attr reference target not found: gitlab.Gitlab.group_access_tokens
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/gl_objects/issues.rst:239: WARNING: py:attr reference target not found: gitlab.issues_statistics
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/gl_objects/notes.rst:19: WARNING: py:attr reference target not found: gitlab.v4.objects.GroupEpic.notes
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/gl_objects/personal_access_tokens.rst:11: WARNING: py:class reference target not found: gitlab.v4.objects.PersonalAcessTokenManager
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/gl_objects/personal_access_tokens.rst:14: WARNING: py:class reference target not found: gitlab.v4.objects.UserPersonalAcessTokenManager
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/gl_objects/personal_access_tokens.rst:15: WARNING: py:attr reference target not found: gitlab.Gitlab.User.personal_access_tokens
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/gl_objects/project_access_tokens.rst:14: WARNING: py:attr reference target not found: gitlab.Gitlab.project_access_tokens
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/gl_objects/repository_tags.rst:12: WARNING: py:attr reference target not found: gitlab.v4.objects.Repository.tags
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/gl_objects/snippets.rst:11: WARNING: py:class reference target not found: gitlab.v4.objects.SnipptManager
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/gl_objects/todos.rst:10: WARNING: py:class reference target not found: gitlab.objects.Todo
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/gl_objects/todos.rst:11: WARNING: py:class reference target not found: gitlab.objects.TodoManager
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/gl_objects/users.rst:219: WARNING: py:attr reference target not found: gitlab.Gitlab.user
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/../gitlab/client.py:docstring of gitlab.client.Gitlab:: WARNING: py:class reference target not found: requests.sessions.Session
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/../gitlab/client.py:docstring of gitlab.client.Gitlab:: WARNING: py:class reference target not found: requests.sessions.Session
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/../gitlab/client.py:docstring of gitlab.client.Gitlab.from_config:: WARNING: py:class reference target not found: config_files
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/../gitlab/client.py:docstring of gitlab.client.Gitlab.http_delete:: WARNING: py:class reference target not found: requests.models.Response
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/../gitlab/client.py:docstring of gitlab.client.Gitlab.http_delete:: WARNING: py:class reference target not found: requests.models.Response
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/../gitlab/client.py:docstring of gitlab.client.Gitlab.http_get:: WARNING: py:class reference target not found: requests.models.Response
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/../gitlab/client.py:docstring of gitlab.client.Gitlab.http_get:: WARNING: py:class reference target not found: requests.models.Response
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/../gitlab/client.py:docstring of gitlab.client.Gitlab.http_post:: WARNING: py:class reference target not found: requests.models.Response
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/../gitlab/client.py:docstring of gitlab.client.Gitlab.http_post:: WARNING: py:class reference target not found: requests.models.Response
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/../gitlab/client.py:docstring of gitlab.client.Gitlab.http_put:: WARNING: py:class reference target not found: requests.models.Response
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/../gitlab/client.py:docstring of gitlab.client.Gitlab.http_put:: WARNING: py:class reference target not found: requests.models.Response
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/../gitlab/client.py:docstring of gitlab.client.Gitlab.http_request:: WARNING: py:class reference target not found: requests.models.Response
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/../gitlab/client.py:docstring of gitlab.client.Gitlab.http_request:: WARNING: py:class reference target not found: requests.models.Response
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/../gitlab/client.py:docstring of gitlab.client.Gitlab.set_license:: WARNING: py:exc reference target not found: GitlabPostError
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/../gitlab/v4/objects/packages.py:docstring of gitlab.v4.objects.packages.GenericPackageManager.upload:: WARNING: py:class reference target not found: pathlib.Path
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/../gitlab/v4/objects/packages.py:docstring of gitlab.v4.objects.packages.GenericPackageManager.upload:: WARNING: py:class reference target not found: pathlib.Path
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/../gitlab/v4/objects/groups.py:docstring of gitlab.v4.objects.groups.GroupManager.import_group:: WARNING: py:obj reference target not found: typing.BinaryIO
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/../gitlab/v4/objects/groups.py:docstring of gitlab.v4.objects.groups.GroupManager.import_group:: WARNING: py:class reference target not found: requests.models.Response
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/../gitlab/v4/objects/groups.py:docstring of gitlab.v4.objects.groups.GroupManager.import_group:: WARNING: py:obj reference target not found: typing.BinaryIO
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/../gitlab/v4/objects/groups.py:docstring of gitlab.v4.objects.groups.GroupManager.import_group:: WARNING: py:class reference target not found: requests.models.Response
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/../gitlab/v4/objects/projects.py:docstring of gitlab.v4.objects.Project.groups:: WARNING: py:class reference target not found: gitlab.v4.objects.projects.ProjectGroupManager
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/../gitlab/v4/objects/projects.py:docstring of gitlab.v4.objects.projects.Project.languages:: WARNING: py:class reference target not found: requests.models.Response
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/../gitlab/v4/objects/projects.py:docstring of gitlab.v4.objects.projects.Project.languages:: WARNING: py:class reference target not found: requests.models.Response
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/../gitlab/v4/objects/projects.py:docstring of gitlab.v4.objects.projects.Project.transfer:: WARNING: py:class reference target not found: project
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/../gitlab/v4/objects/artifacts.py:docstring of gitlab.v4.objects.artifacts.ProjectArtifactManager.download:: WARNING: py:class reference target not found: are not
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/../gitlab/v4/objects/commits.py:docstring of gitlab.v4.objects.commits.ProjectCommit.revert:: WARNING: py:class reference target not found: requests.models.Response
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/../gitlab/v4/objects/commits.py:docstring of gitlab.v4.objects.commits.ProjectCommit.revert:: WARNING: py:class reference target not found: requests.models.Response
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/../gitlab/v4/objects/commits.py:docstring of gitlab.v4.objects.commits.ProjectCommit.signature:: WARNING: py:class reference target not found: requests.models.Response
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/../gitlab/v4/objects/commits.py:docstring of gitlab.v4.objects.commits.ProjectCommit.signature:: WARNING: py:class reference target not found: requests.models.Response
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/../gitlab/v4/objects/environments.py:docstring of gitlab.v4.objects.environments.ProjectEnvironment.stop:: WARNING: py:class reference target not found: requests.models.Response
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/../gitlab/v4/objects/environments.py:docstring of gitlab.v4.objects.environments.ProjectEnvironment.stop:: WARNING: py:class reference target not found: requests.models.Response
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/../gitlab/v4/objects/issues.py:docstring of gitlab.v4.objects.issues.ProjectIssue.closed_by:: WARNING: py:exc reference target not found: GitlabGetErrot
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/../gitlab/v4/objects/issues.py:docstring of gitlab.v4.objects.issues.ProjectIssue.related_merge_requests:: WARNING: py:exc reference target not found: GitlabGetErrot
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/../gitlab/v4/objects/deploy_keys.py:docstring of gitlab.v4.objects.deploy_keys.ProjectKeyManager.enable:: WARNING: py:class reference target not found: requests.models.Response
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/../gitlab/v4/objects/deploy_keys.py:docstring of gitlab.v4.objects.deploy_keys.ProjectKeyManager.enable:: WARNING: py:class reference target not found: requests.models.Response
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/../gitlab/v4/objects/labels.py:docstring of gitlab.v4.objects.labels.ProjectLabel:1: WARNING: py:class reference target not found: gitlab.mixins.PromoteMixin
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/../gitlab/v4/objects/projects.py:docstring of gitlab.v4.objects.projects.ProjectManager.import_bitbucket_server:: WARNING: py:class reference target not found: requests.models.Response
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/../gitlab/v4/objects/projects.py:docstring of gitlab.v4.objects.projects.ProjectManager.import_bitbucket_server:: WARNING: py:class reference target not found: requests.models.Response
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/../gitlab/v4/objects/projects.py:docstring of gitlab.v4.objects.projects.ProjectManager.import_github:: WARNING: py:class reference target not found: requests.models.Response
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/../gitlab/v4/objects/projects.py:docstring of gitlab.v4.objects.projects.ProjectManager.import_github:: WARNING: py:class reference target not found: requests.models.Response
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/../gitlab/v4/objects/projects.py:docstring of gitlab.v4.objects.projects.ProjectManager.import_project:: WARNING: py:class reference target not found: requests.models.Response
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/../gitlab/v4/objects/projects.py:docstring of gitlab.v4.objects.projects.ProjectManager.import_project:: WARNING: py:class reference target not found: requests.models.Response
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/../gitlab/v4/objects/merge_requests.py:docstring of gitlab.v4.objects.merge_requests.ProjectMergeRequest.changes:: WARNING: py:class reference target not found: requests.models.Response
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/../gitlab/v4/objects/merge_requests.py:docstring of gitlab.v4.objects.merge_requests.ProjectMergeRequest.changes:: WARNING: py:class reference target not found: requests.models.Response
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/../gitlab/v4/objects/merge_requests.py:docstring of gitlab.v4.objects.merge_requests.ProjectMergeRequest.merge_ref:: WARNING: py:class reference target not found: requests.models.Response
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/../gitlab/v4/objects/merge_requests.py:docstring of gitlab.v4.objects.merge_requests.ProjectMergeRequest.merge_ref:: WARNING: py:class reference target not found: requests.models.Response
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/../gitlab/v4/objects/merge_requests.py:docstring of gitlab.v4.objects.merge_requests.ProjectMergeRequest.rebase:: WARNING: py:class reference target not found: requests.models.Response
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/../gitlab/v4/objects/merge_requests.py:docstring of gitlab.v4.objects.merge_requests.ProjectMergeRequest.rebase:: WARNING: py:class reference target not found: requests.models.Response
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/../gitlab/v4/objects/milestones.py:docstring of gitlab.v4.objects.milestones.ProjectMilestone:1: WARNING: py:class reference target not found: gitlab.mixins.PromoteMixin
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/../gitlab/v4/objects/pipelines.py:docstring of gitlab.v4.objects.pipelines.ProjectPipeline.cancel:: WARNING: py:class reference target not found: requests.models.Response
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/../gitlab/v4/objects/pipelines.py:docstring of gitlab.v4.objects.pipelines.ProjectPipeline.cancel:: WARNING: py:class reference target not found: requests.models.Response
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/../gitlab/v4/objects/pipelines.py:docstring of gitlab.v4.objects.pipelines.ProjectPipeline.retry:: WARNING: py:class reference target not found: requests.models.Response
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/../gitlab/v4/objects/pipelines.py:docstring of gitlab.v4.objects.pipelines.ProjectPipeline.retry:: WARNING: py:class reference target not found: requests.models.Response
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/../gitlab/v4/objects/repositories.py:docstring of gitlab.v4.objects.repositories.RepositoryMixin.repository_blob:: WARNING: py:class reference target not found: requests.models.Response
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/../gitlab/v4/objects/repositories.py:docstring of gitlab.v4.objects.repositories.RepositoryMixin.repository_blob:: WARNING: py:class reference target not found: requests.models.Response
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/../gitlab/v4/objects/repositories.py:docstring of gitlab.v4.objects.repositories.RepositoryMixin.repository_compare:: WARNING: py:class reference target not found: requests.models.Response
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/../gitlab/v4/objects/repositories.py:docstring of gitlab.v4.objects.repositories.RepositoryMixin.repository_compare:: WARNING: py:class reference target not found: requests.models.Response
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/../gitlab/v4/objects/repositories.py:docstring of gitlab.v4.objects.repositories.RepositoryMixin.update_submodule:: WARNING: py:class reference target not found: requests.models.Response
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/../gitlab/v4/objects/repositories.py:docstring of gitlab.v4.objects.repositories.RepositoryMixin.update_submodule:: WARNING: py:exc reference target not found: GitlabPutError
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/../gitlab/v4/objects/repositories.py:docstring of gitlab.v4.objects.repositories.RepositoryMixin.update_submodule:: WARNING: py:class reference target not found: requests.models.Response
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/../gitlab/v4/objects/sidekiq.py:docstring of gitlab.v4.objects.sidekiq.SidekiqManager.compound_metrics:: WARNING: py:class reference target not found: requests.models.Response
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/../gitlab/v4/objects/sidekiq.py:docstring of gitlab.v4.objects.sidekiq.SidekiqManager.compound_metrics:: WARNING: py:class reference target not found: requests.models.Response
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/../gitlab/v4/objects/sidekiq.py:docstring of gitlab.v4.objects.sidekiq.SidekiqManager.job_stats:: WARNING: py:class reference target not found: requests.models.Response
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/../gitlab/v4/objects/sidekiq.py:docstring of gitlab.v4.objects.sidekiq.SidekiqManager.job_stats:: WARNING: py:class reference target not found: requests.models.Response
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/../gitlab/v4/objects/sidekiq.py:docstring of gitlab.v4.objects.sidekiq.SidekiqManager.process_metrics:: WARNING: py:class reference target not found: requests.models.Response
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/../gitlab/v4/objects/sidekiq.py:docstring of gitlab.v4.objects.sidekiq.SidekiqManager.process_metrics:: WARNING: py:class reference target not found: requests.models.Response
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/../gitlab/v4/objects/sidekiq.py:docstring of gitlab.v4.objects.sidekiq.SidekiqManager.queue_metrics:: WARNING: py:class reference target not found: requests.models.Response
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/../gitlab/v4/objects/sidekiq.py:docstring of gitlab.v4.objects.sidekiq.SidekiqManager.queue_metrics:: WARNING: py:class reference target not found: requests.models.Response
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/../gitlab/v4/objects/users.py:docstring of gitlab.v4.objects.users.User.activate:: WARNING: py:class reference target not found: requests.models.Response
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/../gitlab/v4/objects/users.py:docstring of gitlab.v4.objects.users.User.activate:: WARNING: py:class reference target not found: requests.models.Response
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/../gitlab/v4/objects/users.py:docstring of gitlab.v4.objects.users.User.block:: WARNING: py:class reference target not found: requests.models.Response
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/../gitlab/v4/objects/users.py:docstring of gitlab.v4.objects.users.User.block:: WARNING: py:class reference target not found: requests.models.Response
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/../gitlab/v4/objects/users.py:docstring of gitlab.v4.objects.users.User.deactivate:: WARNING: py:class reference target not found: requests.models.Response
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/../gitlab/v4/objects/users.py:docstring of gitlab.v4.objects.users.User.deactivate:: WARNING: py:class reference target not found: requests.models.Response
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/../gitlab/v4/objects/users.py:docstring of gitlab.v4.objects.users.User.follow:: WARNING: py:class reference target not found: requests.models.Response
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/../gitlab/v4/objects/users.py:docstring of gitlab.v4.objects.users.User.follow:: WARNING: py:class reference target not found: requests.models.Response
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/../gitlab/v4/objects/users.py:docstring of gitlab.v4.objects.User.followers_users:: WARNING: py:class reference target not found: UserFollowersManager
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/../gitlab/v4/objects/users.py:docstring of gitlab.v4.objects.User.following_users:: WARNING: py:class reference target not found: UserFollowingManager
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/../gitlab/v4/objects/users.py:docstring of gitlab.v4.objects.users.User.unblock:: WARNING: py:class reference target not found: requests.models.Response
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/../gitlab/v4/objects/users.py:docstring of gitlab.v4.objects.users.User.unblock:: WARNING: py:class reference target not found: requests.models.Response
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/../gitlab/v4/objects/users.py:docstring of gitlab.v4.objects.users.User.unfollow:: WARNING: py:class reference target not found: requests.models.Response
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/../gitlab/v4/objects/users.py:docstring of gitlab.v4.objects.users.User.unfollow:: WARNING: py:class reference target not found: requests.models.Response
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/../gitlab/cli.py:docstring of gitlab.cli.docs:: WARNING: py:class reference target not found: argparse.ArgumentParser
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/../gitlab/cli.py:docstring of gitlab.cli.docs:: WARNING: py:class reference target not found: argparse.ArgumentParser
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/../gitlab/cli.py:docstring of gitlab.cli.register_custom_action:: WARNING: py:class reference target not found: gitlab.cli.__F
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/../gitlab/cli.py:docstring of gitlab.cli.register_custom_action:: WARNING: py:class reference target not found: gitlab.cli.__F
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/../gitlab/cli.py:docstring of gitlab.cli.register_custom_action:: WARNING: py:class reference target not found: gitlab.cli.__F
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/../gitlab/cli.py:docstring of gitlab.cli.register_custom_action:: WARNING: py:class reference target not found: gitlab.cli.__F
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/../gitlab/cli.py:docstring of gitlab.cli.what_to_cls:: WARNING: py:class reference target not found: module
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/../gitlab/cli.py:docstring of gitlab.cli.what_to_cls:: WARNING: py:class reference target not found: module
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/../gitlab/exceptions.py:docstring of gitlab.exceptions.on_http_error:: WARNING: py:class reference target not found: gitlab.exceptions.__F
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/../gitlab/exceptions.py:docstring of gitlab.exceptions.on_http_error:: WARNING: py:class reference target not found: gitlab.exceptions.__F
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/../gitlab/exceptions.py:docstring of gitlab.exceptions.on_http_error:: WARNING: py:class reference target not found: The exception type to raise -- must inherit from
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/../gitlab/exceptions.py:docstring of gitlab.exceptions.on_http_error:: WARNING: py:class reference target not found: gitlab.exceptions.__F
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/../gitlab/exceptions.py:docstring of gitlab.exceptions.on_http_error:: WARNING: py:class reference target not found: gitlab.exceptions.__F
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/../gitlab/utils.py:docstring of gitlab.utils.response_content:: WARNING: py:class reference target not found: requests.models.Response
/home/tkloczko/rpmbuild/BUILD/python-gitlab-3.4.0/docs/../gitlab/utils.py:docstring of gitlab.utils.response_content:: WARNING: py:class reference target not found: requests.models.Response
done
build succeeded, 112 warnings.
```

**Comment by nejch** [comment]
Thanks @kloczek, this is not a new issue with 3.4.0, correct? I'm looking at the diff and can't see how it would be introduced in 3.4.0.

**Comment by kloczek** [comment]
Probably this is nothing new.
Recently I've changed methoid of generate documentation as I've moved away from setuptools<>sphix integration straight to `sphinx-build` command I've started using it wih `-n` so this is why I'm reporting on massive scale that kind of issues 😄

**Comment by nejch** [comment]
Ok, thanks :smile:  I was confused or a bit because 3.4.0 was just released this morning and it looked like the issue was a reaction for that :) I'll take the time to look at this at some point later.

**Comment by amimas** [comment]
Is this issue still relevant or valid? I couldn't reproduce it when I ran `tox -e docs` command - even after I commented out `nitpick_ignore_regex = [(r"py:.*", r".*")]` in the `docs/conf.py` file, which was added via #2088

**Comment by amimas** [comment]
You're right @nejch . I see those warnings now if I comment out that config line. Not sure why I didn't get them earlier.

**Comment by amimas** [comment]
I was poking into it a bit. If I remove the `build` directory where the generated site is stored, and then run `tox -e docs` again, then the warnings don't appear. But if the `build` directory is removed, then the output shows the warnings.

One change I can think of is to update the `tox` config so that previously built output is removed. That should make it so there's no "intermittent" issue and also will align closely with the CI process. I'll try to look into the warnings later if I can, but what do you think of this suggestion @nejch ?

**Comment by nejch** [comment]
Thanks for looking into it @amimas. If the `build` warnings are false positives because of some artifacts that shouldn't be scanned, then that makes sense to me! But we should check if that's the case. There might also be an ignore config options in that case, then.

**Comment by amimas** [comment]
Not sure if they are "false positive". Looks like the impact of the warning is that the generated docs site ends up with not linking. For example, here's the last 2 reported warnings:

```
/github.com/python-gitlab/docs/gl_objects/wikis.rst:67: WARNING: py:attr reference target not found: gitlab.v4.objects.ProjectWiki.upload [ref.attr]
/github.com/python-gitlab/docs/gl_objects/wikis.rst:68: WARNING: py:attr reference target not found: gitlab.v4.objects.GrouptWiki.upload [ref.attr]
```

If we browse the relevant areas in the docs, notice they didn't generate a link. One of them is because of a typo (i.e. Group**t**Wiki instead of GroupWiki). But the other one is not a typo and yet sphinx can't seem to resolve the reference. Most of the warnings are like this.

![Image](https://github.com/user-attachments/assets/28afc196-4b8b-4c2c-92ee-11453ca8972e)

Found a similar discussion over here: https://discuss.python.org/t/broken-references-in-sphinx-docs/19463

It's more of a quality of docs issue. Unfortunately I can't seem to figure out the root cause. Not well versed in sphinx. Tried GitHub's copilot which ends up suggesting things that are already set/configured.

**Comment by amimas** [comment]
@nejch - Finally I was able to find a way to resolve these.

Let's focus on one individual warning for simplicity. Hopefully all the warnings and fixes are similar.

```
/github.com/python-gitlab/docs/gl_objects/wikis.rst:67: WARNING: py:attr reference target not found: gitlab.v4.objects.ProjectWiki.upload [ref.attr]
```

The above warning is referring to following line: https://github.com/python-gitlab/python-gitlab/blob/055557efe9de9d4ab7b8237f7de7e033a6b02cd9/docs/gl_objects/wikis.rst?plain=1#L67

Sphinx can't resolve that reference. The `upload` method is defined in the `mixins.py` under `UploadMixin` class, which the `ProjectWiki` class inherits from.

https://github.com/python-gitlab/python-gitlab/blob/055557efe9de9d4ab7b8237f7de7e033a6b02cd9/gitlab/mixins.py#L1003-L1010

There are 2 ways to resolve it.

Option 1:

Change the above mentioned line to:

```diff
-   + :attr:`gitlab.v4.objects.ProjectWiki.upload`
+   + :meth:`gitlab.mixins.UploadMixin.upload`
```

Option 2:

Add a docstring in the `ProjectWiki` class giving a reference to the `upload` method from the `UploadMixin` class.

```diff
class ProjectWiki(SaveMixin, ObjectDeleteMixin, UploadMixin, RESTObject):
+    """
+    .. py:attribute:: upload
+
+        Upload files to wiki. Inherited from :class:`~gitlab.mixins.UploadMixin`.
+   """
    _id_attr = "slug"
    _repr_attr = "slug"
    _upload_path = "/projects/{project_id}/wikis/attachments"
```

With the above change, sphinx is now able to resolve the reference and in the docs it generates links like this:

![Image](https://github.com/user-attachments/assets/357d4961-6314-4fb8-a175-16cf8c9908ec)

The `ProjectWiki` class now has `upload` method listed with a reference to `UploadMixin` class:

![Image](https://github.com/user-attachments/assets/0bf88fe6-d0ce-4356-8f19-a03eeec12d44)

Clicking on the `UploadMixin` link will take us directly to the docs for that class containing the `upload` method:

![Image](https://github.com/user-attachments/assets/8f61b4b9-9686-48ba-b22d-3702252be7f6)

If we choose go with Option 1, then the link in the `wikis.html` will have a direct link to the `upload` method under the `UploadMixin` class.

Which option do you think is most appropriate? What I like about Option 2 is that the `ProjectWiki` class's docs shows that it has a method named `upload`. However, the class also inherits other mixins. So, going this route means maybe those should also be documented the same way. It's a little extra work, but probably is more complete?

What's your thought?
